File: src/main/java/com/kishlaly/ta/analyze/testing/sl/StopLossFixedKeltnerBottom.java
package com.kishlaly.ta.analyze.testing.sl;

import com.kishlaly.ta.model.SymbolData;
import com.kishlaly.ta.model.indicators.Keltner;
import com.kishlaly.ta.utils.IndicatorUtils;

import java.util.List;

/**
 * SL is chosen at the lower point of the Keltner channel, which exists in the signal quote
 */
public class StopLossFixedKeltnerBottom extends StopLossStrategy {

    public StopLossFixedKeltnerBottom() {
        super(null, false);
    }

    @Override
    public double calculate(SymbolData data, int signalIndex) {
        List<Keltner> keltnerChannels = IndicatorUtils.buildKeltnerChannels(data.symbol, data.quotes);
        return keltnerChannels.get(signalIndex).getLow();
    }

    @Override
    public String toString() {
        return "SL [Fixed] Keltner 100% bottom";
    }
}

File: src/main/kotlin/com/kishlaly/ta/analyze/testing/sl/StopLossFixedKeltnerBottom.kt
package com.kishlaly.ta.analyze.testing.sl

import com.kishlaly.ta.model.SymbolData
import com.kishlaly.ta.utils.IndicatorUtils

class StopLossFixedKeltnerBottom : StopLossStrategy() {

    override fun calculate(data: SymbolData, signalIndex: Int): Double {
        return IndicatorUtils.buildKeltnerChannels(data.symbol, data.quotes)[signalIndex].low
    }

    override fun toString() = "SL [Fixed] Keltner 100% bottom"

}